File: android/src/main/kotlin/com/solusibejo/analytics_debugger/AnalyticsDebuggerMethods.kt
package com.solusibejo.analytics_debugger

import android.app.Activity
import com.solusibejo.analytics_debugger.consts.Arguments
import com.solusibejo.analytics_debugger.debugger.DebuggerManager
import com.solusibejo.analytics_debugger.debugger.DebuggerMode
import com.solusibejo.analytics_debugger.model.EventProperty
import io.flutter.plugin.common.MethodCall

object AnalyticsDebuggerMethods {
    fun show(
        activity: Activity,
        debugger: DebuggerManager,
        call: MethodCall,
    ){
        val mode = call.argument<Boolean>(Arguments.debuggerMode)!!
        val isSystemWide = call.argument<Boolean?>(Arguments.systemWide) ?: false
        var type = DebuggerMode.bubble

        if(!mode) type = DebuggerMode.bar

        if(isSystemWide){
            debugger.showDebugger(activity, type, true)
        }
        else {
            debugger.showDebugger(activity, type)
        }
    }

    fun hide(
        activity: Activity,
        debugger: DebuggerManager,
    ) {
        debugger.hideDebugger(activity)
    }

    fun send(
        debugger: DebuggerManager,
        call: MethodCall,
    ){
        val name = call.argument<String>(Arguments.name)!!
        val values = call.argument<Map<String, Any>?>(Arguments.values)!!

        val events: ArrayList<EventProperty> = ArrayList()
        for ((key, value) in values){
            events.add(EventProperty(key, value.toString()))
        }

        debugger.publishEvent(
            System.currentTimeMillis(),
            name,
            events
        )
    }
}

File: ios/Classes/AnalyticsDebuggerMethods.swift
//
//  AnalyticsDebuggerMethods.swift
//  analytics_debugger
//
//  Created by Chandra Abdul Fattah on 12/02/23.
//

import Foundation
import Flutter
import AnalyticsDebugger

class AnalyticsDebuggerMethods {
    static func show(
        debugger: AnalyticsDebugger,
        call: FlutterMethodCall
    ){
        let args = call.arguments as! [String : Any]

        let mode = args[Arguments.debuggerMode] as! Bool
        
        if(mode){
            debugger.showBubble()
        }
        else {
            debugger.showBarDebugger()
        }
    }
    
    static func hide(
        debugger: AnalyticsDebugger
    ){
        debugger.hide()
    }
    
    static func send(
        debugger: AnalyticsDebugger,
        call: FlutterMethodCall
    ){
        let args = call.arguments as! [String : Any]
        
        let name = args[Arguments.name] as! String
        let params = args[Arguments.values] as? [String: Any]
        
        let events = params.map { param in
            var values = Array<DebuggerProp>()
            for (key, value) in param {
                values.append(DebuggerProp(id: key, withName: key, withValue: value))
            }
            return values
        }
    
        debugger.publishEvent(
            name,
            withTimestamp: NSNumber(value: NSDate().timeIntervalSince1970),
            withProperties: events ?? [DebuggerProp]()
        )
    }
}
